config.go

```diff
@@ -0,0 +1,329 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"go.uber.org/config/internal/merge"
+	"go.uber.org/config/internal/unreachable"
+
+	"go.uber.org/multierr"
+	"golang.org/x/text/transform"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const _separator = "."
+
+// YAML is a provider that reads from one or more YAML sources. Many aspects
+// of the resulting provider's behavior can be altered by passing functional
+// options.
+//
+// By default, the YAML provider attempts to proactively catch common mistakes
+// by enabling gopkg.in/yaml.v2's strict mode. See the package-level
+// documentation on strict unmarshalling for details.
+//
+// When populating Go structs, values produced by the YAML provider correctly
+// handle all struct tags supported by gopkg.in/yaml.v2. See
+// https://godoc.org/gopkg.in/yaml.v2#Marshal for details.
+type YAML struct {
+	name     string
+	raw      []byte
+	contents interface{}
+	strict   bool
+}
+
+// NewYAML constructs a YAML provider. See the various YAMLOptions for
+// available tweaks to the default behavior.
+func NewYAML(options ...YAMLOption) (y *YAML, retErr error) {
+	cfg := &config{
+		strict: true,
+		name:   "YAML",
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, cfg.close())
+	}()
+	for _, o := range options {
+		o.apply(cfg)
+	}
+
+	if cfg.err != nil {
+		return nil, fmt.Errorf("error applying options: %v", cfg.err)
+	}
+
+	// On construction, go through a full merge-serialize-deserialize cycle to
+	// catch any duplicated keys as early as possible (in strict mode). It also
+	// strips comments, which stops us from attempting environment variable
+	// expansion.
+	merged, err := merge.YAML(cfg.sources, cfg.strict)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't merge YAML sources: %v", err)
+	}
+
+	if cfg.lookup != nil {
+		exp, err := ioutil.ReadAll(transform.NewReader(merged, newExpandTransformer(cfg.lookup)))
+		if err != nil {
+			return nil, fmt.Errorf("couldn't expand environment: %v", err)
+		}
+		merged = bytes.NewBuffer(exp)
+	}
+
+	raw := merged.Bytes() // helpful for WithDefault later on
```

Comment:
Unfortunately, we can't do that because we're using the buffer as a reader a few lines down. Once all the contents of the buffer are consumed via that API, `.Bytes()` returns an empty slice.

We could capture the slice here and use `Unmarshal`/`UnmarshalStrict` below, but I thought that was a little uglier.

Reply:
Ah got it, I wasn't aware. Thanks for clarifying! I prefer this implementation over your suggestion.